tools/ml-dataprep/dataprep.py

```diff
@@ -0,0 +1,249 @@
+
+    def _build_destination_table(self, timestamp, suffix):
+        """Create the name of a destination table based on the source table.
+
+        The name of the destination table  is obtained by appending to the name of the source
+        table name a timestamp and a suffix.
+
+        Args:
+            timestamp: An int representing the timestamp to append.
+            suffix: A string representing the suffix to append.
+
+        Returns:
+            A string containing the destination table name.
+            A string containing the fully qualified BigQuery name of the destination table.
+        """
+        table_id = '{}_{}_{}'.format(self._source_table,
+                                     suffix,
+                                     timestamp)
+        table_uri = self._bq_client.build_table_uri(self._destination_project,
+                                                    self._destination_dataset,
+                                                    table_id)
+        return table_id, table_uri
+
+    def _calculate_dataset_sizes(self):
+        """Calculate the size of the training and validation datasets.
+
+        Calculates the total number of lines in the source table as well as the split index at
+        which to separate the training data and the validation data based on the value of the
+        self._split_ratio parameter.
+
+        Returns:
+            An int containing the total number of lines in the source table.
+            An int containing the split index at which separate training and validation data.
+        """
+        total_lines = self._bq_client.count_lines_in_table(
+            self._source_table_uri)
+        split_index = int(self._split_ratio * total_lines)
+        return total_lines, split_index
+
+    def _create_temp_table(self, timestamp, total_lines):
+        """Save in a temporary table the data in the source table shuffled.
+
+        Creates a copy of the data from the source table and saves it in a temporary table with
+        an additional RAND() column used for shuffling the data.
+
+        Args:
+            timestamp: A long containing the timestamp to append in the name of the temporary
+            table.
+            total_lines: An int containing the total number of lines in the source table.
+
+        Returns:
+            A string containing the fully qualified BigQuery name of the temporary table.
+        """
+        _, temp_table_uri = self._build_destination_table(
+            timestamp, TEMP_TABLE_SUFFIX)
+        query = queries.QUERY_TEMP_DATA_TEMPLATE.format(temp_table_uri,
+                                                        self._columns,
+                                                        self._target_columns_shuffle,
+                                                        self._source_table_uri,
+                                                        total_lines)
+        self._bq_client.run_query(query)
```

Comment:
BQ automatically creates temp tables for your queries. You should be able to run a query that assigns random indices and export results from the temp table that BQ will create. The benefit is you don't need to cleanup the temp table manually

Reply:
I agree that we could use the temp table created by BigQuery after just running the query. However, if I'm not mistaken, I see the risk of running into the "Response too large error" when the compressed result exceeds 128 MB. Given that we are dealing with ML datasets this is highly probable. Maybe the temporary table as still been created in spite of the error and we could use it, but I don't see this as a viable mechanism. We would assume that some things went fine in spite of the error thrown. The reason for deliberately creating the temporary table (and subsequently having to clean it up) was, among others, to avoid such errors.

If you see a workaround we're more than interested.